README.md
# [q1tsim](https://github.com/Q1tBV/q1tsim/)

[![Build Status](https://travis-ci.org/Q1tBV/q1tsim.svg?branch=master)](https://travis-ci.org/Q1tBV/q1tsim)
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
[![crates.io](http://meritbadge.herokuapp.com/q1tsim)](https://crates.io/crates/q1tsim)
[![Released API docs](https://docs.rs/q1tsim/badge.svg)](https://docs.rs/q1tsim)

A simple, efficient, quantum computer simulator.

Overview
========

q1tsim is a simulator library for a quantum computer, written in Rust. Its goal
is to be an easy to use, efficient simulator for the development and testing of
quantum algorithms.

Features
========
* Easy implementation and simulation of quantum circuits
* Supports the creation of arbitrary quantum gates
* Most common quantum gates already included
* Measurement in `X`, `Y`, or `Z` basis
* Possibility of measurement without affecting the quantum state
* Creation of histograms of measurement results over multiple runs
* Operations conditional on classical values
* Export of circuits to Open QASM and c-QASM for running your programs on other computers or simulators
* Export of circuits to LaTeX, for drawing pictures of your circuit
* Efficient simulation of stabilizer circuits

Usage
=====
To use q1tsim in your Rust application, add the following to your `Cargo.toml` file:

```toml
[dependencies]
q1tsim = "0.5"
```

As an example, here is a 3-qubit quantum Fourier transform of the |000⟩ quantum
state:
```
extern crate q1tsim;

use q1tsim::{circuit, gates};

fn main()
{
    // The number of times this circuit is evaluated
    let nr_runs = 8192;

    // Create a quantum circuit with 3 quantum bits and 3 classical (measurement)
    // bits. The circuit starts by default with all quantum bits in the |0⟩ state,
    // so in this case |000⟩.
    let mut circuit = circuit::Circuit::new(3, 3);

    // Set up a 3-qubit quantum Fourier transform
    // There is no predefined method on Circuit that implements a controlled
    // `S` or `T` gate, so we use the `add_gate()` method for those.
    circuit.h(2);
    circuit.add_gate(gates::CS::new(), &[1, 2]);
    circuit.add_gate(gates::CT::new(), &[0, 2]);
    circuit.h(1);
    circuit.add_gate(gates::CS::new(), &[0, 1]);
    circuit.h(0);
    circuit.add_gate(gates::Swap::new(), &[0, 2]);

    // Measure all quantum bits in the Pauli `Z` basis
    circuit.measure_all(&[0, 1, 2]);

    // Actually calculate the resulting quantum state and perform the measurements,
    // averaging over `nr_runs` runs.
    circuit.execute(nr_runs);

    // And print the results.
    let hist = circuit.histogram_string().unwrap();
    for (bits, count) in hist
    {
        println!("{}: {}", bits, count);
    }
}
```
The result should be a more or less equal distribution over the eight possible
states (000, 001, ..., 111).

Read the complete API documentation on [docs.rs](https://docs.rs/q1tsim).


CHANGELOG.md
# [0.5.0] 2019-09-13

- Added reference parameters. This allows one to change the parameters to a gate
between executions of a circuit without having to create a new circuit (and also
preserving the quantum state).
- Added a python interface to add the option of using `q1tsim` simulator from Python.
- Update the dependency on the `rand` crate to version 0.7. This fixes a situation
in drawing from a binomial distribution being very slow when the success chance was
close to zero.
- Small bug fixes and code reorganization.

# [0.4.0] 2019-05-29

Much more efficient simulation of stabilizer circuits. Stabilizer circuits consist
of Clifford gates only: gates that can be described entirely in terms of `H`, `S`
and `CX` gates. These circuits can be represented more efficiently, scaling
quadratically in the number of qubits instead of exponentially as when using the
standard coefficient vector representation (see [ArXiv:9807006](https://arxiv.org/abs/quant-ph/9807006)
for details). The gates included with `q1tsim` that can be simulated in this
representation are `H`, `X`, `Y`, `Z`, `S` and `S`<sup>`†`</sup>,
`V` and `V`<sup>`†`</sup>, `CX`, `CY`, `CZ`, and `Swap`,
as well as `Composite` gates containing only these gates.  With the new backend,
simulation of stabilizer circuits with thousands of qubits is possible.

By default, `q1tsim` will now use the stabilizer representation if possible. If
you wish to use the coefficient vector representation instead when simulating a
stabilizer circuit, use the `Circuit::execute_with()` method.

# [0.3.0] 2019-04-09

- Update code to use Rust 2018 edition.
- Most methods of `Circuit` now return a `Result` with a possible error code.
Checking the result of every call while building a circuit may be tedious. In that
case, use the new `circuit!` macro to build the circuit, which stops at the first
error it encounters, e.g.
```
let circuit = circuit!(2, 2, {
    h(0);
    x(1);
    measure_all(&[0, 1]);
}).expect("Failed to build circuit");
```
- The number of runs to average over is no longer specified in `Circuit::new()`,
but passed in later when calling `Circuit::execute()`.
- A new `Circuit::reexecute()` method was added, to execute an existing circuit
again without resetting the state, i.e. starting in the end state of the previous
execution.
- The `Gate` trait no longer requires you to implement the `cost()` function.
Its application is limited, and sometimes hard to estimate. The function now
returns infinity by default.
- New `peek()` and `peek_all()` methods on `Circuit` that allow you to perform
a measurement without collapsing the quantum state. This is of course not a
physical process, and cannot be reproduced on a quantum computer, but can be
useful for simualting intermediate measurements while still evolving the
quantum state.
- The random number generator used for sampling in measurements is no longer
stored in `QuState`, but instead passed in. Combined with the new `execute_with_rng()`
and reexecute_with_rng()` methods on `Circuit`, this allows users to use their
own RNG for sampling. This can be useful, e.g., to create reproducible simulations.
- The internal representation of QuState was simplified to use a single matrix
of quantum states instead of a loose collection of individual states. This allows
us to perform operations on all states at once. Calculations with heavy branching
(due to measurements), should run much faster now.
- Several updates and fixes in export to various formats.
- New `s()` and `sdg` methods on `Circuit` for quickly adding `S` or `S`<sup>`†`</sup>
phase gates to a circuit.
- Added doubly controlled rotation gates `CCRX`, `CCRY`, and `CCRZ`.


# [0.2.0] 2019-02-27

- Much better test coverage
- Export of circuits to `LaTeX` code, using the `qcircuit` package.
- Arithmetic expressions, and the constant `pi` can now be used in describing
gate parameters in `Composite::from_string()`.
- Measurements off all bits can now be performed in a different basis than just
the Pauli `Z` basis.
- Several small performance improvements
- Fixed wrong bit order of the control register in classically controlled
operations. The bit order is now consistent with the histograms from
measurements (i.e. first control bit is now least significant).
- Several bug fixes in the export to OpenQasm en c-Qasm.
- Fixed parameters being passed in the wrong order by `Circuit::u3()`.
- Removed `Gate::expanded_matrix()`. It was used nowhere, and it can easily be
reimplemented more generally by operating with the gate on a matrix. Use the
identity matrix to obtain the effect of the original function.

# [0.1.0] 2019-01-30

Initial version


src/lib.rs
// Copyright 2019 Q1t BV
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! A simple, efficient, quantum computer simulator.
//!
//! Overview
//! ========
//!
//! q1tsim is a simulator library for a quantum computer, written in Rust. Its goal
//! is to be an easy to use, efficient simulator for the development and testing of
//! quantum algorithms.
//!
//! Features
//! ========
//! * Easy implementation and simulation of quantum circuits
//! * Supports the creation of arbitrary quantum gates
//! * Most common quantum gates already included
//! * Measurement in `X`, `Y`, or `Z` basis
//! * Possibility of measurement without affecting the quantum state
//! * Creation of histograms of measurement results over multiple runs
//! * Operations conditional on classical values
//! * Export of circuits to Open QASM and c-QASM for running your programs on other computers or simulators
//! * Export of circuits to LaTeX, for drawing pictures of your circuit
//! * Efficient simulation of stabilizer circuits
//!
//! Usage
//! =====
//! To use q1tsim in your Rust application, add the following to your `Cargo.toml` file:
//!
//! ```toml
//! [dependencies]
//! q1tsim = "0.3"
//! ```
//!
//! As an example, here is a 3-qubit quantum Fourier transform of the |000⟩ quantum
//! state:
//! ```
//! use q1tsim::circuit::Circuit;
//! use q1tsim::gates::{CS, CT, Swap};
//!
//! fn main()
//! {
//!     // The number of times this circuit is evaluated
//!     let nr_runs = 8192;
//!
//!     // Create a quantum circuit with 3 quantum bits and 3 classical (measurement)
//!     // bits. The circuit starts by default with all quantum bits in the |0⟩
//!     // state, so in this case |000⟩.
//!     let mut circuit = Circuit::new(3, 3);
//!
//!     // Set up a 3-qubit quantum Fourier transform
//!     // There is no predefined method on Circuit that implements a controlled
//!     // `S` or `T` gate, so we use the `add_gate()` method for those.
//!     circuit.h(2);
//!     circuit.add_gate(CS::new(), &[1, 2]);
//!     circuit.add_gate(CT::new(), &[0, 2]);
//!     circuit.h(1);
//!     circuit.add_gate(CS::new(), &[0, 1]);
//!     circuit.h(0);
//!     circuit.add_gate(Swap::new(), &[0, 2]);
//!
//!     // Measure all quantum bits in the Pauli `Z` basis
//!     circuit.measure_all(&[0, 1, 2]);
//!
//!     // Actually calculate the resulting quantum state, and perform the
//!     // measurements, averaging over `nr_runs` runs.
//!     circuit.execute(nr_runs);
//!
//!     // And print the results.
//!     let hist = circuit.histogram_string().unwrap();
//!     for (bits, count) in hist.iter()
//!     {
//!         println!("{}: {}", bits, count);
//!     }
//! }
//! ```
//! The result should be a more or less equal distribution over the eight possible
//! states (000, 001, ..., 111).
//!
//! Creating a circuit
//! ==================
//! Struct [Circuit](circuit/struct.Circuit.html) is the main structure used
//! in creating a quantum program. The basic layout of a program to create
//! a quantum circuit, execute it, and collect the results, is as follows:
//! ```
//! use q1tsim::circuit::Circuit;
//!
//! // Create a new circuit with `nr_qbits` quantum bits and `nr_cbits`
//! // classical bits
//! let nr_qbits = 2;
//! let nr_cbits = 2;
//! let mut circuit = Circuit::new(nr_qbits, nr_cbits);
//!
//! // Add operations on the circuit. In this case, a Hadamard transform on the
//! // first bit, followwed by a CNOT gate with the first bit as control and
//! // the second bit as target.
//! circuit.h(0);
//! circuit.cx(0, 1);
//!
//! // Add a measurement of the resulting quantum state. This measures the first
//! // qbit into classical bit 0, and the second qbit into classical bit 1.
//! circuit.measure_all(&[0, 1]);
//!
//! // Now execute the circuit, averaging measurements over `nr_runs` runs
//! // of the circuit.
//! let nr_runs = 1024;
//! circuit.execute(nr_runs);
//!
//! // And finally collect the results. The `histogram_vec()` method returns a
//! // vector with at each index `i` the number if times the measurement returned
//! // `i` in the classical register.
//! let hist = circuit.histogram_vec();
//! ```
//! Since version 0.3, many of the methods on `Circuit` will return a `Result`,
//! possibly containing an error code (e.g. if invalid bit numbers are used).
//! Checking the result of each modification of the circuit quickly becomes
//! tedious, so the [circuit](macro.circuit.html) macro was added that can make
//! multiple method calls and immediately returns on the first error encountered
//! (or returns `Ok(())` if all calls were successful). With this, the previous
//! program can be written as
//! ```rust
//! # #[macro_use] extern crate q1tsim; fn main() {
//! let nr_qbits = 2;
//! let nr_cbits = 2;
//! let mut circuit = circuit!(nr_qbits, nr_cbits, {
//!     h(0);
//!     cx(0, 1);
//!     measure_all(&[0, 1]);
//! }).expect("Failed to build circuit");
//!
//! let nr_runs = 1024;
//! circuit.execute(nr_runs);
//! let hist = circuit.histogram_vec();
//! # }
//!```
//!
//! Custom gates
//! ============
//! Using the [Circuit::add_gate()](circuit/struct.Circuit.html#method.add_gate)
//! method, arbitrary gates can be added to a circuit. You can define your own
//! custom gates by implementing the [Gate](gates/trait.Gate.html) trait. To
//! implement this trait, the type should implement at least the
//! [description()](gates/trait.Gate.html#tymethod.description),
//! [nr_affected_bits()](gates/trait.Gate.html#tymethod.nr_affected_bits),
//! and [matrix()](gates/trait.Gate.html#tymethod.matrix) methods. The
//! `description()` method should return a short textual identifier or label
//! for the gate, while `nr_affected_bits()` returns the number of qubits on
//! which the gate operates. The `matrix()` method should return a matrix of size
//! `2`<sup>`n`</sup>`×2`<sup>`n`</sup>, where `n` is the number of affected bits,
//! that describes the unitary transformation that the gate implements. An example
//! of a simple custom gate that rotates the `|01⟩` and `|10⟩` components of
//! a pair of qubits, is given below:
//! ```
//! use ndarray::array;
//! use q1tsim::ExportGate;
//!
//! #[derive(ExportGate)]
//! struct Mix
//! {
//!    alpha: f64
//! }
//!
//! impl q1tsim::gates::Gate for Mix
//! {
//!     fn description(&self) -> &str { "M" }
//!     fn nr_affected_bits(&self) -> usize { 2 }
//!     fn matrix(&self) -> q1tsim::cmatrix::CMatrix
//!     {
//!         let o = q1tsim::cmatrix::COMPLEX_ONE;
//!         let z = q1tsim::cmatrix::COMPLEX_ZERO;
//!         let c = self.alpha.cos() * o;
//!         let s = self.alpha.sin() * o;
//!         array![
//!             [o, z,  z, z],
//!             [z, c, -s, z],
//!             [z, s,  c, z],
//!             [z, z,  z, o]
//!         ]
//!     }
//! }
//! ```
//! Types implementing the `Gate` trait may optionally also implement the
//! [apply()](gates/trait.Gate.html#method.apply) family of methods if a more
//! optimal implementation than simply multiplying by its associated matrix can
//! be found.
//!
//! Exporting gates and circuits
//! ============================
//! The discerning reader may have notices the `#[derive(ExportGate)]` statement
//! on the custom gate in the listing above. This makes the type use the default
//! implementations of the export functions for a gate. Currently, there are
//! three traits for exporting a gate:
//! - [OpenQasm](export/trait.OpenQasm.html) for exporting a gate to OpenQasm code.
//! - [CQasm](export/trait.CQasm.html) for exporting a gate to c-Qasm code.
//! - [Latex](export/trait.Latex.html) for exporting a gate to LaTeX.
//!
//! You can use the default implementation for each of these traits by deriving
//! them, e.g.
//! ```
//! use q1tsim::OpenQasm;
//! use q1tsim::gates::Gate;
//!
//! #[derive(OpenQasm)]
//! struct MySpecialGate {}
//!
//! impl Gate for MySpecialGate {
//!     /* ... */
//!     # fn description(&self) -> &str { "" }
//!     # fn nr_affected_bits(&self) -> usize { 0 }
//!     # fn matrix(&self) -> q1tsim::cmatrix::CMatrix { q1tsim::cmatrix::CMatrix::zeros((0,0)) }
//! }
//! ```
//! The default implementations for OpenQasm and CQasm simply return an error,
//! since there is no way [^no_qasm] to know how to encode a custom gate
//! in these formats. The default implementation for the LaTeX export simply draws
//! a rectangular box with the gate description inside. As seen before, if you
//! want to use default definitions for all export traits, derive from `ExportGate`.
//!
//! Note that to use a gate type in a circuit, it must be exportable, so an
//! implementation for the export traits must be defined for your custom type,
//! either through deriving or by providing your own implementation.
//!
//! [^no_qasm]: No reasonable way at least. Technically, we could take the matrix
//! for the gate, decompose it into primitive gates, and export the corresponding
//! code.
//!
//! Stabilizer circuits
//! ===================
//! Stabilizer circuits are circuits that can be expressed entirely in terms of
//! the Clifford gates `H`, `S`, and `CX`, and qubit measurements. Since version
//! 0.4.0, `q1tsim` can simulate these circuits much more efficiently than
//! general circuits. If you have a custom gate type that can be represented
//! in terms of Clifford gates, and wish to use it with the stabilizer backend,
//! you should override the default implementations of the
//! [is_stabilizer()](gates/trait.Gate.html#method.is_stabilizer) and
//! [conjugate()](gates/trait.Gate.html#method.conjugate) methods. As an example,
//! the implementation for a hypothetical `HX` gate that first performs a Hadamard
//! transform, followed by an `X` gate, could look like
//! ```
//! use q1tsim::stabilizer::PauliOp;
//! use q1tsim::gates::Gate;
//!
//! struct HX {}
//!
//! impl Gate for HX {
//!     # fn description(&self) -> &str { "" }
//!     # fn nr_affected_bits(&self) -> usize { 0 }
//!     # fn matrix(&self) -> q1tsim::cmatrix::CMatrix { q1tsim::cmatrix::CMatrix::zeros((0,0)) }
//!     fn is_stabilizer(&self) -> bool
//!     {
//!         true
//!     }
//!
//!     fn conjugate(&self, ops: &mut [PauliOp]) -> q1tsim::error::Result<bool>
//!     {
//!         let (op, sign) = match ops[0]
//!             {
//!                 PauliOp::I => (PauliOp::I, false),
//!                 PauliOp::Z => (PauliOp::X, true),
//!                 PauliOp::X => (PauliOp::Z, false),
//!                 PauliOp::Y => (PauliOp::Y, false)
//!             };
//!         ops[0] = op;
//!         Ok(sign)
//!     }
//!
//!     /* ... */
//! }
//! ```

#[macro_use] extern crate ndarray;
#[cfg(test)] #[macro_use] extern crate matches;

#[macro_use] pub mod cmatrix;
#[macro_use] pub mod gates;
pub mod arithmetic;
pub mod circuit;
pub mod error;
pub mod ffi;
pub mod export;
pub mod expression;
pub mod permutation;
pub mod qustate;
pub mod vectorstate;
pub mod stabilizer;

mod idhash;
mod support;
#[cfg(test)] mod stats;

pub use q1tsim_derive::*;
